vscode/microsoft-kiota/src/openApiTreeProvider.ts

```diff
@@ -214,33 +214,39 @@ export class OpenApiTreeProvider implements vscode.TreeDataProvider<OpenApiTreeN
             }
         }
     }
-    getCollapsedState(hasChildren: boolean): vscode.TreeItemCollapsibleState {
-        return !hasChildren ?
-                vscode.TreeItemCollapsibleState.None :
-                (this.tokenizedFilter.length === 0 ?
-                    vscode.TreeItemCollapsibleState.Collapsed : 
-                    vscode.TreeItemCollapsibleState.Expanded);
+    getCollapsedState(node: KiotaOpenApiNode): vscode.TreeItemCollapsibleState {
+        const hasChildren = node.children.length > 0;
+        if (!hasChildren) {
+            return vscode.TreeItemCollapsibleState.None;
+        }
+        if (this.tokenizedFilter.length === 0) {
+            return vscode.TreeItemCollapsibleState.Expanded;
+        }
+
+        if (this.tokenizedFilter.some(x => node.children.some(c => c.segment.includes(x) || this.getCollapsedState(c) === vscode.TreeItemCollapsibleState.Expanded))) {
+            return vscode.TreeItemCollapsibleState.Expanded;
+        }
+        return vscode.TreeItemCollapsibleState.Collapsed;
     }
-    getTreeNodeFromKiotaNode(node: KiotaOpenApiNode): OpenApiTreeNode {
+    getTreeNodeFromKiotaNode(node: KiotaOpenApiNode, collapsibleStateOverride: vscode.TreeItemCollapsibleState | undefined = undefined): OpenApiTreeNode {
         return new OpenApiTreeNode(
-            node.path, 
+            node.path,
             node.segment === pathSeparator && this.apiTitle ? pathSeparator + " (" + this.apiTitle + ")" : node.segment,
             node.selected ?? false,
-            this.getCollapsedState(node.children.length > 0),
+            collapsibleStateOverride ?? this.getCollapsedState(node),
             node.isOperation ?? false,
             node.children.map(x => this.getTreeNodeFromKiotaNode(x)),
             node.documentationUrl
         );
     }
-    getChildren(element?: OpenApiTreeNode): OpenApiTreeNode[] {
+    getChildren(element?: OpenApiTreeNode): vscode.ProviderResult<OpenApiTreeNode[]> {
         if (!this.rawRootNode) {
             return [];
         }
         if (element) {
             return element.children.filter(x => x.isNodeVisible(this.tokenizedFilter));
         } else {
-            const result = this.getTreeNodeFromKiotaNode(this.rawRootNode);
-            result.collapsibleState = vscode.TreeItemCollapsibleState.Expanded;
+            const result = this.getTreeNodeFromKiotaNode(this.rawRootNode, vscode.TreeItemCollapsibleState.Expanded);
```

Comment:
> Thanks for looking into this. Yes I've pulled my hair a lot on the tree view and the visibility.

Well, I can only say that I'm sorry, and I feel your pain... :(

> From what I can remember it seems that in some cases the refresh triggers a partial re-render for some reason.

Yeah, it very much seems to be the case.

> So I'd suggest you set breaking points to double check whether the getCollaspedState method is called as expected, and returns the expected value for nodes.

Yes, I did so actually, and all the data returned seems to be correct - see screenshots - and for reference:
```
	export enum TreeItemCollapsibleState {
		None = 0,
		Collapsed = 1,
		Expanded = 2
	}
```

![image](https://github.com/microsoft/kiota/assets/26791344/975da453-0391-4049-a1b0-dea150932a79)

![image](https://github.com/microsoft/kiota/assets/26791344/1ade8ef2-bea5-4e1d-9ce3-ff44f83b7fda)

![image](https://github.com/microsoft/kiota/assets/26791344/6d5b42b0-e32a-43d2-ba84-0f59a2d97488)

![image](https://github.com/microsoft/kiota/assets/26791344/91b6e78e-55c3-4d49-9c70-69ace8ef5f96)

![image](https://github.com/microsoft/kiota/assets/26791344/dae183aa-9027-4da4-945d-b1b365eb4035)


> Additionally, looking [at the documentation](https://code.visualstudio.com/api/extension-guides/tree-view) I can see a difference in implementation which might or not have an impact.

I made the change, it doesn't seem to have any impact.

> L272, the property should be readonly.
> Which means L250 we won't be able to set it externally, which means we need an optional parameter for getCollaspedState to pass an overriding value when needed.

Sure, done.

> I hope all of that helps debug things. Let me know if you have further questions.

Well, I don't really feel like spending too much more time fighting with VS Code internals I'm afraid. :( 
As much as it's annoying that collapsing doesn't work reliably, IMO the changes I proposed make the overall experience better - so I'd be in the camp of merging it as is, and potentially following up with the improvement later. There is also a chance that the problem will be fixed on the VS Code side independently, perhaps?

Reply:
I've taken some time to look into it. apparently this "strange" behavior is by design. Changing a previously displayed node from "Collapsed" to "Expanded" will be ignored by vscode, we'd need to call the reveal API for this to be applied properly.
https://github.com/microsoft/vscode/issues/131955#issuecomment-912366684

The challenge with this API is that it'd require maintaining a state in sync for two trees (raw nodes and tree nodes), which is never good in terms of reliability. The alternative being re-project the raw nodes tree into the tree nodes tree multiple time per filter (which already runs 1 time per character typed), which would be disastrous in terms of performance...

I think we can table that as a separate issue, I have reverted the default expansion state changes as for large APIs having everything expanded by default it far from a great experience.

Let me know if you want to explore the expansion issue further, we can create a separate issue for it.

Reply:
Thanks for looking into it @baywet - I'm happy with how it is now, let's have it merged. :)

Reply:
even better news! I had a lovely chat with @alexr00 (one of the engineers behind the treeview) and she shared details about how the id is handled with regards to the "refresh" of the collapsed state. I was able to fix the behaviour!
Resolving